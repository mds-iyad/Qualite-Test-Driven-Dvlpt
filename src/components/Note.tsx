
export interface Note {
    id: number;
    title: string;
    content: number;
    comment: string;
    creationDate: string;
  }
  
export type NoteState = {
    notes: Note[];
    searchTerm: string;
    sortBy: string;
    sortOrder: 'asc' | 'desc';
    addNote: (note: Note) => void;
    deleteNote: (id: number) => void;
    updateNote: (id: number, updatedNote: Note) => void;
    setSearchTerm: (term: string) => void;
    setSortBy: (value: string) => void;
    setSortOrder: (order: 'asc' | 'desc') => void;
  };