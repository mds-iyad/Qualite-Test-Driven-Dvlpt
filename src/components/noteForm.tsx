import React, { useState, useEffect } from 'react';
import {Note, NoteState} from './Note';
import create from 'zustand';
import { persist, createJSONStorage } from 'zustand/middleware'




export const useNoteStore = create<NoteState>(persist((set) => ({
  notes: [],
  searchTerm: '',
  sortBy: 'creationDate',
  sortOrder: 'desc',
  addNote: (note) => set((state) => {
    return { notes: [...state.notes, note] }
  }),
  deleteNote: (id) => set((state) => ({ notes: state.notes.filter((note) => note.id !== id) })),
  updateNote: (id, updatedNote) =>
    set((state) => ({
      notes: state.notes.map((note) => (note.id === id ? updatedNote : note)),
    })),
  setSearchTerm: (term) => set(() => ({ searchTerm: term })),
  setSortBy: (value) => set(() => ({ sortBy: value })),
  setSortOrder: (order) => set(() => ({ sortOrder: order })),
  reset: () => set(() => ({   notes: [],
    searchTerm: '',
    sortBy: 'creationDate',
    sortOrder: 'desc' })),
}),
{
  name: 'note-storage',
  storage: createJSONStorage( ()=> sessionStorage),
}
));

export const NoteForm: React.FC = () => {
  const [title, setTitle] = useState('');
  const [content, setContent] = useState(0);
  const [comment, setComment] = useState('');

  const addNote = useNoteStore((state) => state.addNote);

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();

    const newNote: Note = {
      id: new Date().getTime(),
      title,
      content,
      comment,
      creationDate: new Date().toLocaleString(),
    };

    addNote(newNote);

    setTitle('');
    setContent(0);
    setComment('');
  };

  console.log(title)

  return (
    <form onSubmit={handleSubmit} data-testid="form">
      <input data-testid="input-title"
        type="text"
        placeholder="Titre"
        value={title}
        onChange={(e) => setTitle(e.target.value)}
        required
      />
      <input data-testid="input-note"
        type="number"
        placeholder="Note"
        value={content}
        onChange={(e) => setContent(e.target.value)}
        required
      ></input>
      <textarea data-testid="input-comment"
        placeholder="Commentaire"
        value={comment}
        onChange={(e) => setComment(e.target.value)}
        required
      ></textarea>
      <button data-testid="add-button" type="submit">Ajouter</button>
    </form>
  );
};