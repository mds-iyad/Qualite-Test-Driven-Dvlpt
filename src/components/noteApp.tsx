import React, { useState, useEffect } from 'react';
import {useNoteStore,NoteForm} from './noteForm';
import { NoteList } from './NoteList';



const NoteApp: React.FC = () => {
  const setSearchTerm = useNoteStore((state) => state.setSearchTerm);
  const setSortBy = useNoteStore((state) => state.setSortBy);
  const setSortOrder = useNoteStore((state) => state.setSortOrder);

  return (
    <div className="note-app">
      <h1>Gestionnaire de notes</h1>
      <NoteForm />
      <input
        type="text"
        data-testid="input-search"
        placeholder="Rechercher..."
        onChange={(e) => setSearchTerm(e.target.value)}
      />
      <select data-testid="sort-by-select" value={useNoteStore((state) => state.sortBy)} onChange={(e) => setSortBy(e.target.value)}>
        <option value="creationDate">Date de création</option>
        <option value="note">Note</option>
      </select>
      <select data-testid="sort-order-select"
        value={useNoteStore((state) => state.sortOrder)}
        onChange={(e) => setSortOrder(e.target.value as 'asc' | 'desc')}
      >
        <option value="desc">Décroissant</option>
        <option value="asc">Croissant</option>
      </select>
      <NoteList />
    </div>
  );
};

export default NoteApp;
