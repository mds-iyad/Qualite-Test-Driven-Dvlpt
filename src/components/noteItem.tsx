import React, { useState, useEffect } from 'react';
import {Note, NoteState} from './Note';
import {useNoteStore,NoteForm} from './noteForm';

export const NoteItem: React.FC<{ note: Note }> = ({ note }) => {
    const deleteNote = useNoteStore((state) => state.deleteNote);
    const updateNote = useNoteStore((state) => state.updateNote);
  
    const [isEditing, setIsEditing] = useState(false);
    const [editedTitle, setEditedTitle] = useState(note.title);
    const [editedContent, setEditedContent] = useState(note.content);
    const [editedComment, setEditedComment] = useState(note.comment);
  
    const handleDelete = () => {
      // if (window.confirm('Êtes-vous sûr de vouloir supprimer cette note ?')) 
      // {
      //   deleteNote(note.id);
      // }
      deleteNote(note.id);
    };
  
    const handleEdit = () => {
      setIsEditing(true);
    };
  
    const handleSave = () => {
      const updatedNote: Note = {
        ...note,
        title: editedTitle,
        content: editedContent,
        comment: editedComment,
      };
      updateNote(note.id, updatedNote);
      setIsEditing(false);
    };
    const getNoteColor = (note: Note): string => {
      const noteLength = note.content;
      if (noteLength < 8) {
        return 'red';
      } else if (noteLength < 10) {
        return 'orange';
      } else if (noteLength < 13) {
        return 'yellow';
      } else {
        return 'green';
      }
    };
  
    return (
      <div className="note-item" style={{ backgroundColor: getNoteColor(note) }}>
        <div className="note-info">
          {isEditing ? (
            <input
              type="text"
              value={editedTitle}
              onChange={(e) => setEditedTitle(e.target.value)}
            />
          ) : (
            <h3 data-testid="note-title">{note.title}</h3>
          )}
          <p>{note.creationDate}</p>
          {isEditing ? (
            <textarea value={editedComment} onChange={(e) => setEditedComment(e.target.value)} />
          ) : (
            <p>{note.comment.substring(0, 100)}</p>
          )}
          {isEditing ? (
            <input value={editedContent} onChange={(e) => setEditedContent(e.target.value)} />
          ) : (
            <p>{note.content}</p>
          )}
        </div>
        <div className="note-actions">
          {isEditing ? (
            <button onClick={handleSave}>Enregistrer</button>
          ) : (
            <>
              <button onClick={handleEdit}>Modifier</button>
              <button onClick={handleDelete}>Supprimer</button>
            </>
          )}
        </div>
      </div>
    );
  };