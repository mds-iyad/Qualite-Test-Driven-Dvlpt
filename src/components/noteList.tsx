import React, { useState, useEffect } from 'react';
import {useNoteStore} from './noteForm';
import {NoteItem} from './noteItem'



export const NoteList: React.FC = () => {
    const notes = useNoteStore((state) => state.notes);
    const searchTerm = useNoteStore((state) => state.searchTerm);
    const sortBy = useNoteStore((state) => state.sortBy);
    const sortOrder = useNoteStore((state) => state.sortOrder);
  
    const filteredNotes = notes.filter(
      (note) =>
        note.title.toLowerCase().includes(searchTerm.toLowerCase()) ||
        note.content.toString().toLowerCase().includes(searchTerm.toLowerCase()) ||
        note.comment.toLowerCase().includes(searchTerm.toLowerCase())
    );
  
    const sortedNotes = filteredNotes.sort((a, b) => {
      const factor = sortOrder === 'asc' ? 1 : -1;
  
      if (sortBy === 'creationDate') {
        return factor * (new Date(a.creationDate).getTime() - new Date(b.creationDate).getTime());
      } else if (sortBy === 'note') {
        return factor * (a.content - b.content);
      }
  
      return 0;
    });
  
    return (
      <div className="note-list">
        {sortedNotes.map((note) => (
          <NoteItem data-testid="note-title" key={note.id} note={note} />
        ))}
      </div>
    );
  };