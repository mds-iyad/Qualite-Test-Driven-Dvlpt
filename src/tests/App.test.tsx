import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import App from '../App';
import { test, expect, describe, vi } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import NoteApp from '../components/NoteApp';
import { useNoteStore } from '../components/noteForm'

import axios from 'axios';

//vi.mock('axios');



describe('NoteApp', () => {

	beforeEach(() => {
		const { reset } = useNoteStore.getState();
		reset();
	});


	// test('couleur type', async () => {

	// 	render(<NoteApp />)


	// })

  test('affiche la liste des notes existantes',async  () => {
    render(<NoteApp />);
		// Ajoute 2 notes
		await userEvent.type(screen.getByTestId('input-title'), 'title 1')
		await userEvent.type(screen.getByTestId('input-note'), "16")
		await userEvent.type(screen.getByTestId('input-comment'), 'Note 1')
		await fireEvent.submit(screen.getByTestId('form'))
	
		await userEvent.type(screen.getByTestId('input-title'), 'title 2')
		await userEvent.type(screen.getByTestId('input-note'), "18")
		await userEvent.type(screen.getByTestId('input-comment'), 'Note 2')
		await fireEvent.submit(screen.getByTestId('form'))
	
    // Vérifie si les notes sont affichées
    expect(screen.getByText('Note 1')).toBeInTheDocument();
    expect(screen.getByText('Note 2')).toBeInTheDocument();
  });



  test('modifie une note existante',async () => {
    render(<NoteApp />);
	await userEvent.type(screen.getByTestId('input-title'), 'Titre de la note 1')
	await userEvent.type(screen.getByTestId('input-note'), "16")
	await userEvent.type(screen.getByTestId('input-comment'), 'Commentaire de la note 1')
	await fireEvent.submit(screen.getByTestId('form'))

    // Clique sur une note pour la modifier
    await fireEvent.click(screen.getByText('Modifier'));

    // Modifie le titre, le contenu et le commentaire de la note
	const title = screen.getByDisplayValue('Titre de la note 1');
    await userEvent.clear(title);
    await userEvent.type(title, 'Nouveau titre');
    /*await userEvent.clear(screen.getByDisplayValue("16"));
    await userEvent.type(screen.getByDisplayValue("16"), "17");
    await userEvent.clear(screen.getByDisplayValue('Commentaire de la note 1'));
    await userEvent.type(screen.getByDisplayValue('Commentaire de la note 1'), 'Nouveau commentaire');*/
    await fireEvent.click(screen.getByText('Enregistrer'));

	console.log(screen.debug())


    // Vérifie si la note a été modifiée
    expect(screen.getByText('Nouveau titre')).toBeInTheDocument();
    //expect(screen.getByText(17)).toBeInTheDocument();
    //expect(screen.getByText('Nouveau commentaire')).toBeInTheDocument();
  });

  test('supprime une note existante', async () => {
    render(<NoteApp />);

    await userEvent.type(screen.getByTestId('input-title'), 'Titre de la note 1')
    await userEvent.type(screen.getByTestId('input-note'), "16")
    await userEvent.type(screen.getByTestId('input-comment'), 'Commentaire de la note 1')
    await fireEvent.submit(screen.getByTestId('form'))

    // Clique sur une note pour la supprimer
    fireEvent.click(screen.getByText('Supprimer'));

    // Confirme la suppression de la note
    //fireEvent.click(screen.getByText('Ok'));

    // Vérifie si la note a été supprimée
    expect(screen.queryByText('Titre de la note 1')).not.toBeInTheDocument();
  });

  test.only('recherche des notes', async () => {
    render(<NoteApp />);

    await userEvent.type(screen.getByTestId('input-title'), 'Titre de la note 1')
    await userEvent.type(screen.getByTestId('input-note'), "16")
    await userEvent.type(screen.getByTestId('input-comment'), 'Commentaire de la note 1')
    await fireEvent.submit(screen.getByTestId('form'))

    await userEvent.type(screen.getByTestId('input-title'), 'Titre de la note 2')
    await userEvent.type(screen.getByTestId('input-note'), "16")
    await userEvent.type(screen.getByTestId('input-comment'), 'Commentaire de la note 2')
    await fireEvent.submit(screen.getByTestId('form'))

    await userEvent.type(screen.getByTestId('input-title'), 'Titre de la note 3')
    await userEvent.type(screen.getByTestId('input-note'), "16")
    await userEvent.type(screen.getByTestId('input-comment'), 'Commentaire de la note 3')
    await fireEvent.submit(screen.getByTestId('form'))

    // Saisit un terme de recherche
    userEvent.type(screen.getByTestId("input-search"), 'note 2');

    // Vérifie si seules les notes correspondantes au terme de recherche sont affichées
    expect(screen.getByText('Titre de la note 2')).toBeInTheDocument();
    expect(screen.queryByText('Titre de la note 1')).not.toBeInTheDocument();
    expect(screen.queryByText('Titre de la note 3')).not.toBeInTheDocument();
  });

  test('trie les notes par date de création', async () => {
    render(<NoteApp />);

    await userEvent.type(screen.getByTestId('input-title'), 'Titre de la note 1')
    await userEvent.type(screen.getByTestId('input-note'), "16")
    await userEvent.type(screen.getByTestId('input-comment'), 'Commentaire de la note 1')
    await fireEvent.submit(screen.getByTestId('form'))

    await userEvent.type(screen.getByTestId('input-title'), 'Titre de la note 2')
    await userEvent.type(screen.getByTestId('input-note'), "16")
    await userEvent.type(screen.getByTestId('input-comment'), 'Commentaire de la note 2')
    await fireEvent.submit(screen.getByTestId('form'))

    await userEvent.type(screen.getByTestId('input-title'), 'Titre de la note 3')
    await userEvent.type(screen.getByTestId('input-note'), "16")
    await userEvent.type(screen.getByTestId('input-comment'), 'Commentaire de la note 3')
    await fireEvent.submit(screen.getByTestId('form'))

    // Sélectionne l'option de tri par date de création (ordre croissant)
    fireEvent.change(screen.getByTestId('sort-by-select'), { target: { value: 'creationDate' } });
    fireEvent.change(screen.getByTestId('sort-order-select'), { target: { value: 'asc' } });

    // Vérifie si les notes sont triées par date de création (ordre croissant)
    const noteTitles = screen.getAllByTestId('note-title').map((element) => element.textContent);
    expect(noteTitles).toEqual(['Titre de la note 1', 'Titre de la note 2', 'Titre de la note 3']);
  });

  test('trie les notes par note',async ()  => {
    render(<NoteApp />);

    await userEvent.type(screen.getByTestId('input-title'), 'Titre de la note 1')
    await userEvent.type(screen.getByTestId('input-note'), "16")
    await userEvent.type(screen.getByTestId('input-comment'), 'Commentaire de la note 1')
    await fireEvent.submit(screen.getByTestId('form'))

    await userEvent.type(screen.getByTestId('input-title'), 'Titre de la note 2')
    await userEvent.type(screen.getByTestId('input-note'), "16")
    await userEvent.type(screen.getByTestId('input-comment'), 'Commentaire de la note 2')
    await fireEvent.submit(screen.getByTestId('form'))

    await userEvent.type(screen.getByTestId('input-title'), 'Titre de la note 3')
    await userEvent.type(screen.getByTestId('input-note'), "16")
    await userEvent.type(screen.getByTestId('input-comment'), 'Commentaire de la note 3')
    await fireEvent.submit(screen.getByTestId('form'))

    // Sélectionne l'option de tri par note (ordre décroissant)
    fireEvent.change(screen.getByTestId('sort-by-select'), { target: { value: 'note' } });
    fireEvent.change(screen.getByTestId('sort-order-select'), { target: { value: 'desc' } });

    // Vérifie si les notes sont triées par note (ordre décroissant)
    const noteTitles = screen.getAllByTestId('note-title').map((element) => element.textContent);
    expect(noteTitles).toEqual(['Titre de la note 3', 'Titre de la note 2', 'Titre de la note 1']);
  });


});

