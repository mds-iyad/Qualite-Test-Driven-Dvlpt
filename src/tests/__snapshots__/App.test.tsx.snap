// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NoteApp > Test can add a note 1`] = `
[
  {
    "comment": "comment test",
    "content": 14,
    "creationDate": "02/06/2023, 15:47:05",
    "id": 1685713625401,
    "title": "comment test",
  },
]
`;
